src/Runner.php

```diff
@@ -2,87 +2,95 @@
      */
-    public function __construct(array $options = [])
+    public function __construct(array $options = ['application' => null, 'config' => null,])
     {
-        $this->commands_directory = __DIR__ . '/Commands';
-        $this->application = new Terminus();
-        $this->configureApplication(new Config($options));
+        $this->application = $options['application'];
+        $this->config = $options['config'];
+
+        $commands_directory = __DIR__ . '/Commands';
+        $top_namespace = 'Pantheon\Terminus\Commands';
+        $commands = $this->getCommands(['path' => $commands_directory, 'namespace' => $top_namespace,]);
+        $this->runner = new RoboRunner($commands);
     }
 
     /**
      * Runs the instantiated Terminus application
+     *
+     * @param string[] $arguments Argv from the command line
+     * @return integer The exiting status code of the application
      */
-    public function run()
+    public function run(array $arguments = [])
     {
-        $this->application->run();
+        $status_code = $this->runner->execute($arguments, null, 'Terminus', $this->config->get('version'));
+        return $status_code;
     }
 
     /**
-     * Adds command files to the application
+     * Discovers command classes using CommandFileDiscovery
      *
-     * @param Config $config A Terminus configuration object
-     * @return void
+     * @param string[] $options Elements as follow
+     *        string path      The full path to the directory to search for commands
+     *        string namespace The full namespace associated with given the command directory
+     * @return TerminusCommand[] An array of TerminusCommand instances
      */
-    private function configureApplication(Config $config)
+    private function getCommands(array $options = ['path' => null, 'namespace' => null,])
     {
-        $command_factory = new AnnotatedCommandFactory();
         $discovery = new CommandFileDiscovery();
-        $discovery->setSearchPattern('*Command.php');
-        $subdirectories = array_merge(
-            $this->locateSubdirectories($this->commands_directory, $this->top_namespace),
-            $this->locateSubdirectories($config->get('plugins_dir'))
-        );
-        foreach ($subdirectories as $subdirectory) {
-            $command_files = $discovery->discover($subdirectory['path'], $subdirectory['namespace']);
-            foreach ($command_files as $command_class) {
-                $reflection_class = new \ReflectionClass($command_class);
-                if (!$reflection_class->isAbstract()) {
-                    $command_list = $command_factory->createCommandsFromClass(new $command_class($config));
-                    foreach ($command_list as $command) {
-                        $this->application->add($command);
-                    }
-                }
-            }
-        }
+        $discovery->setSearchPattern('*Command.php')->setSearchLocations(['.']);
```

Comment:
I'll give you a new CommandFileDiscovery that will work right if you change this to `setSearchLocations([]);`. The problem with `.` as a search location is it makes the command discovery object search each location twice, because it always searches the base directory atm.